src/Interpreters/MySQL/InterpretersMySQLDDLQuery.cpp

```diff
@@ -92,22 +92,38 @@ static NamesAndTypesList getColumnsList(const ASTExpressionList * columns_defini
         }
 
         ASTPtr data_type = declare_column->data_type;
+        auto * data_type_function = data_type->as<ASTFunction>();
 
-        if (is_unsigned)
+        if (data_type_function)
         {
-            auto * data_type_function = data_type->as<ASTFunction>();
+            String type_name_upper = Poco::toUpper(data_type_function->name);
 
-            if (data_type_function)
+            if (is_unsigned)
             {
-                String type_name_upper = Poco::toUpper(data_type_function->name);
-
                 /// For example(in MySQL): CREATE TABLE test(column_name INT NOT NULL ... UNSIGNED)
-                if (type_name_upper.find("INT") != std::string::npos && !endsWith(type_name_upper, "SIGNED")
+                if (type_name_upper.find("INT") != String::npos && !endsWith(type_name_upper, "SIGNED")
                     && !endsWith(type_name_upper, "UNSIGNED"))
                     data_type_function->name = type_name_upper + " UNSIGNED";
             }
-        }
 
+            /// Transforms MySQL ENUM's list of strings to ClickHouse string-integer pairs
+            /// For example ENUM('a', 'b', 'c') -> ENUM('a'=1, 'b'=2, 'c'=3)
+            if (type_name_upper.find("ENUM") != String::npos)
+            {
+                Int64 i = 0;
```

Comment:
MySQL Supports values from 1 to 65535 here, while ClickHouse uses a signed 16-bit integer. This means we need to map MySQL values from 32768 to 65535 to negative numbers, -32768 to -1. This also requires changes in the replication code of course.

Reply:
But if we go from Enum8 to Enum16 with this mapping, we will mix up our values. What would be good approach to prevent it? 
Also, maybe, we can shift everything backwards on -32768 (like, 0-> -32768, 1-> -32767...)? Then the integer values will remain in the same order as they do in MySQL, and that might be useful for some queries. Also will make easier to convert Enum8 to Enum16 (just substract some number, so we start from -32768 and not -128), as well as allowing us to easily change ClickHouse values to MySQL values in queries (just add 32768 to them)

Reply:
Or perhaps if we don't want to change an entire column of values when going from Enum8 to Enum16, as well as we don't want to have negative numbers representing our Enum - we can use only half of Enum8 values [1,127], if we reach 128 - we switch to Enum16 (preserving all the values in columns), and if we reach 32768 - we go to negative numbers, as you suggested. That way ClickHouse enum would almost never be different to MySQL enum (I imagine that enums for more then 32767 values are not really used).